### DISCUSSION TITLE
Access violation when building in shared library

### DISCUSSION THREAD

**Discussion by DominicPicard** [fnc12/sqlite_orm#1267] [Q&A]
Hi,

I have been using sqlite_orm in a library in MSVC C++. Everything is going except when I build a .dll instead of a .lib

I get an access violation exception, this is the callstack : 
![image](https://github.com/fnc12/sqlite_orm/assets/23101453/5d6b6844-9505-4dde-98e6-27aeceac4db9)

I searched on google for this specific error but I could not find anything.

I am using Conan to retrieve sqlite3 and sqlite_orm. I use sqlite 3.43.2

I was wondering if anyone has ever seen this weird problem ?

I am starting to panic 😟

**Comment by trueqbit** [comment]
Did you create sqlite3 with debug symbols (pdb)? I suggest you investigate the context around the crashing code path a bit more. Also try a build without optimizations.

**Reply by DominicPicard** [reply]
Hi,

Thank you for your quick reply !

I debugged a bit and I see that the address for the global variable **sqlite3GlobalConfig** changes...

That global variable holds the mutex function pointer which becomes null when it is recreated : 
![image](https://github.com/fnc12/sqlite_orm/assets/23101453/8db4f0e2-6680-4bdd-b015-6c2fc83791a6)

The callstack is the following:
![image](https://github.com/fnc12/sqlite_orm/assets/23101453/6ec7c084-feec-4b99-bc32-bc5fe2cf1299)

And the exception happens here : 
![image](https://github.com/fnc12/sqlite_orm/assets/23101453/31991bbf-6a54-4e72-8175-bff9e816c33d)


My work around is to call **sqlite3_initialize()** here :
![image](https://github.com/fnc12/sqlite_orm/assets/23101453/0a288122-b22a-43a2-90e4-9a63d434339a)

I think my issue is more related to C/C++, why does that global variable gets recreated ? Do I do something wrong when I include headers/libraries ?

**Comment by trueqbit** [comment]
Usually, those kind of things rather happen in multithreaded applications. Does this possibly apply to you?
The way connections are currently handled in sqlite_orm requires that you use open_forever() when using multiple threads.

**Reply by DominicPicard** [reply]
I am not sure what the issue is. I had the same problem when using SQLiteCpp. I changed my design the restrict the database access to the library and not also access the database directly in my unit test application.
